app/models/concerns/taggable_terms.rb

```diff
@@ -0,0 +1,77 @@
+require 'set'
+
+# Apply tags to a Taggable record based on configured attributes matching terms
+# mapped to tags get applied when there is a match.
+#
+# Example:
+#
+#   Record.taggable_terms = { body: { /foo/ => 'foo' } }
+#   r = Record.new
+#   r.tagged?('foo')
+#   # => false
+#
+#   r.update(body: 'foo bar baz')
+#   r.tagged?('foo')
+#   # => true
+module TaggableTerms
+  extend ActiveSupport::Concern
+
+  included do
+    cattr_accessor :taggable_terms, default: {}
+    before_save :update_taggable_terms, if: :taggable_term_attribute_changed?
+  end
+
+  def update_taggable_terms
+    tags_to_add, tags_to_remove = taggable_terms_changed_tags
+    tags_to_add.each { |tag| add_tag_if_not_already_present(tag) }
+    tags_to_remove.each { |tag| remove_tag(tag) }
+  end
+
+  private
+
+  def taggable_term_attribute_changed?
+    changed.include?(taggable_terms.keys)
+  end
+
+  def taggable_terms_changed_tags
+    tags_to_add = Set.new
+    tags_to_remove = Set.new
+
+    taggable_terms_to_tag_to_attr_terms.each do |tag, attr_terms_pairs|
+      tag_str = tag.to_s
+
+      attr_terms_pairs.each do |attr, term|
+        if attribute_matches_taggable_term?(attr, term)
+          tags_to_add << tag_str
+          break
+        end
+      end
+
+      tags_to_remove << tag_str unless tags_to_add.include?(tag_str)
+    end
+
+    [tags_to_add, tags_to_remove]
+  end
+
+  def attribute_matches_taggable_term?(attr, term)
+    read_attribute(attr) =~ Regexp.new(term)
+  end
+
+  # Restructure taggable_terms so that it's more processing friendly
+  #
+  # Before:
+  # => {:body=>{/train/i=>"trains", /bus/i=>"bus", /locomotive/i=>"trains"}}
+  #
+  # After:
+  # => {"trains"=>[[:body, /train/i], [:body, /locomotive/i]],
+  #     "bus"=>[[:body, /bus/i]]}
```

Comment:
> Have you thought of using `Regexp.union` so this would return something like:

I did think about whether it might be worth munging into a single regexp. Are there any performance issues with that past a certain point?

> Also what about if the body contains say "constrain", we could in theory convert the terms into a Regexp (if not already) and wrap between word boundaries EG: `\b...\b`

In real use I'm expecting the regexps to be more like `/my 999 call/`, `/my visa application/`, etc – more phrases than singular words. I think if we do want to match a singular word (`SAR`) for example, then when we configure it we explicitly include things in the regexp that reduces the chances of false positives.